embassy-net-adin1110/src/lib.rs

```diff
@@ -0,0 +1,1249 @@
+
+pub const PHYID: u32 = 0x0283BC91;
+
+/// Error values ADIN1110
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[allow(non_camel_case_types)]
+pub enum AdinError<E> {
+    Spi(E),
+    SENDERROR,
+    READERROR,
+    CRC,
+    PACKET_TOO_BIG,
+    PACKET_TOO_SMALL,
+    MDIO_ACC_TIMEOUT,
+}
+
+pub type AEResult<T, SPIE> = core::result::Result<T, AdinError<SPIE>>;
+pub const MDIO_PHY_ADDR: u8 = 0x01;
+pub const MTU: usize = 1500;
+/// Max SPI/Frame buffer size
+pub const MAX_BUFF: usize = 2048;
+
+const DONT_CARE_BYTE: u8 = 0x00;
+const TURN_AROUND_BYTE: u8 = 0x00;
+
+const FEC_LEN: usize = 4;
+const FRAME_HEADER_LEN: usize = 2;
+const WR_HEADER_LEN: usize = 2;
+
+// P1 = 0x00, P2 = 0x01
+const PORT_ID_BYTE: u8 = 0x00;
+
+pub type Packet = Vec<u8, { MTU + FEC_LEN + WR_HEADER_LEN }>;
+
+/// Type alias for the embassy-net driver for ADIN1110
+pub type Device<'d> = embassy_net_driver_channel::Device<'d, MTU>;
+
+/// Internal state for the embassy-net integration.
+pub struct State<const N_RX: usize, const N_TX: usize> {
+    ch_state: ch::State<MTU, N_RX, N_TX>,
+}
+impl<const N_RX: usize, const N_TX: usize> State<N_RX, N_TX> {
+    /// Create a new `State`.
+    pub const fn new() -> Self {
+        Self {
+            ch_state: ch::State::new(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ADIN1110<SPI> {
+    /// SPI bus
+    spi: SPI,
+    /// Enable CRC on SPI transfer.
+    /// This must match with the hardware pin SPI_CFG0 were 0 = CRC enable, 1 CRC disabled.
+    crc: bool,
+}
+
+// Round size up the N u32;
+pub(crate) fn size_align_u32(size: u32) -> u32 {
+    (size + 3) & 0xFFFF_FFFC
+}
+
+impl<SpiE, SPI> ADIN1110<SPI>
+where
+    SPI: SpiDevice<u8, Error = SpiE>,
+    SpiE: core::fmt::Debug,
+{
+    pub fn new(spi: SPI, crc: bool) -> Self {
+        Self { spi, crc }
+    }
+
+    pub async fn read_reg(&mut self, reg: sr) -> AEResult<u32, SpiE> {
+        let mut tx_buf = Vec::<u8, 16>::new();
+
+        let mut spi_hdr = SpiHeader(0);
+        spi_hdr.set_control(true);
+        spi_hdr.set_addr(reg);
+        let _ = tx_buf.extend_from_slice(spi_hdr.0.to_be_bytes().as_slice());
+
+        if self.crc {
+            // Add CRC for header data
+            let _ = tx_buf.push(crc8(&tx_buf));
+        }
+
+        // Turn around byte, TODO: Unknown that this is.
+        let _ = tx_buf.push(TURN_AROUND_BYTE);
+
+        let mut rx_buf = [0; 5];
+
+        let spi_read_len = if self.crc { rx_buf.len() } else { rx_buf.len() - 1 };
+
+        let mut spi_op = [Operation::Write(&tx_buf), Operation::Read(&mut rx_buf[0..spi_read_len])];
+
+        self.spi.transaction(&mut spi_op).await.map_err(AdinError::Spi)?;
+
+        if self.crc {
+            let crc = crc8(&rx_buf[0..4]);
+            if crc != rx_buf[4] {
+                return Err(AdinError::CRC);
+            }
+        }
+
+        let value = u32::from_be_bytes(rx_buf[0..4].try_into().unwrap());
+
+        #[cfg(feature = "defmt")]
```

Comment:
you can copypaste [`fmt.rs`](https://github.com/embassy-rs/embassy/blob/main/embassy-net-driver-channel/src/fmt.rs) into this crate, it allows you to do `trace!` unconditionally, it takes care of automatically forwarding to `defmt`, `log` or none depending on Cargo features.

if you do, also add `log` as an optional dep.